docs/website/docs/HardWonWisdom.md

```diff
@@ -112,3 +112,123 @@ REQ2: If L≤k (see REQ1) and L<R, we must validate at least L+1 of their header
+
+In general, it would be possible to design a system such that even the final ledger state in an era did not determine when that era would end.
+That'd be an empty safe zone, consisting of zero slots.
+For example, perhaps the first block of an era identifies itself as such simply by setting some field in its header.
+The design of the Ouroboros Praos protocol itself prevents Cardano from doing something like that (TODO did the Byron protocol also? Is that named "Ouroboros Classic"? "Ouroboros BFT"?).
+In particular, a Praos ledger state must determine the leader schedule for some non-empty prefix of the upcoming epochs.
+If only for that reason, then, the Cardano ledger rules do not support abrupt changes to the election rules, such as era changes (at the very least, the leader schedule is slot- and epoch-based while eras can change the slot length and/or the epoch size).
+Such changes must come with sufficient forewarning (spoiler: _**two**_ stability windows) before they take effect and they must take effect at some epoch boundary.
+Thus, every era transition happens as part of an epoch transition, and whether or not an epoch transition includes an era transition is determined by at least one ledger state that precedes the epoch transition by at least some minimum number of slots (spoiler: _**two**_ stability windows).
+
+Warning.
+(Skip this on your first read of this section.)
+That reasoning implies some non-empty suffix of the ledger states in an era will each determine the exact end time of that era.
+It's technically possible a ledger state could also determine the end time of subseqeuent eras, but the Cardano ledger rules do not permit that.
+On `mainnet` Cardano, only blocks in an era can cause that era to end.
+This also implies each era with a known end time on a `mainnet` chain must contain some blocks on that chain.
+On the other hand, the Consensus Layer configuration inputs do permit overriding the era transition rules for the sake of testing, in which case a ledger state could anticipate the end time of several subsequent eras and eras could be empty (eg main use case is skipping from Byron to some later Shelley era ASAP, in which case all the intervening eras have no blocks but also have no slots!).
+
+There is one last notable caveat at this abstract level.
+The Hard Fork Combinator safe zone for each `mainnet` era is exactly the minimum number of slots of forewarning before the end of that era.
+One might naturally expect this is the stability window (see the above entry in this file for why this is `3k/f` in Praos), but it is instead twice that.
+The best resource we have for motivating this decision is [Edsko de Vries' IOG seminar from 2020 June on the HFC and time](https://drive.google.com/file/d/1QIJ-VBlj-txB6K6E7DIEnY5TzaD89qQm/view).
```

Comment:
I think something is being mixed up here; we set the `eraSafeZone` field in `EraParams` (one of which for each era live in `Shape` via `HardForkConsensusConfig`) to $3k/f$ and not $6k/f$ for all Shelley-based eras:

https://github.com/input-output-hk/ouroboros-consensus/blob/e19a58d078b882e5180d9ef591a1f4e9a1987fe0/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs#L134-L140

(for reference: [`SL.computeStabilityWindow`](https://github.com/input-output-hk/cardano-ledger/blob/180271602640bcac1214084b6de61d0468332f00/eras/shelley/impl/src/Cardano/Ledger/Shelley/StabilityWindow.hs#L19-L20))

I also skimmed the linked talk again and couldn't find where "safe zone = two stability windows" is stated.

Maybe the difference is between the number of slots it takes for an proposal to come into effect (which is $6k/f$, and I think is what you are trying to describe here, in particular with the reference to the `PUPP` ledger rule) compared to the number of slots ahead of a ledger state in which an era transition is guaranteed not to occur if it is not already known (cf. which is the definition of "safe zone" in [section 17.3.3 in the report](https://input-output-hk.github.io/ouroboros-network/pdfs/report#subsection.17.3.3))?

Reply:
Hmm. Yes. I think I did the conflation you were talking about.

1) How early does the ledger state determine when the transition will happen, _assuming_ it does happen? At least 6k/f before the epoch transition.

2) How early does the ledger state determine when the transition will happen _and_ that it definitely will happen? At least 3k/f before the epoch transition.

I'll think through this more carefully and rewrite.

Reply:
I rewrote it.

Reply:
I added another couple paragraphs; spelling out the fundamentals of Header-Body Split's requirements on the stability of header validity. Phew!

Reply:
Thanks, that is a very useful thing to explore, I left a few comments.

Also, this might be sufficiently separate from the overall topic to move to its own section, or even put it in [section 24.5 ("Eliminating safe zones") in the report](https://input-output-hk.github.io/ouroboros-network/pdfs/report/report.pdf#section.24.5) (or the succeeding section).